### ISSUE TITLE
Large time delay/white space between subsequent dask.compute() calls

### ISSUE THREAD

**Issue by scottshambaugh** [dask/distributed#7066]
**What happened**:
I use dask as a way to do parallel processing of single functions, and [the repo where I use it](https://github.com/scottshambaugh/monaco/blob/main/src/monaco/mc_sim.py) essentially runs a single-item task graph three times. However, I am seeing huge time delays between consecutive runs of the task graphs.

I have tried to boil down the issue to a minimum viable example which should have essentially no delay due to serialization or surrounding code. The below example mirrors the rough structure of how I use dask in my code, and reproduces the the issue I am seeing. Attached is the html performance report, but in short this the large delay between the two groups of tasks is what I'm looking at:

![image](https://user-images.githubusercontent.com/14363975/192167057-97f406a4-23b1-41f0-8ea9-be24a14757bc.png)

**What you expected to happen**:
There should not be such a large delay between the two groups of tasks.

**Minimal Complete Verifiable Example**:

```python
import dask
from dask.distributed import Client, performance_report
from numpy.random import default_rng

def case_runner(case):
    return round(case)

class Sim:
    def __init__(self, ncases=int(1e3)):
        self.ncases = ncases
        self.rng = default_rng(seed=12345)
        self.cases = self.rng.random(self.ncases)
        self.runcases = []

        self.client = Client()
        self.cluster = self.client.cluster
        print(f'Dask dashboard link: {self.cluster.dashboard_link}')

    def run(self, n):
        with performance_report():
            for _ in range(n):
                runcases = []
                for case in self.cases:
                    case_delayed = dask.delayed(case_runner)(case)
                    runcases.append(case_delayed)
                runcases = dask.compute(*runcases)
                self.runcases = runcases

def main():
    sim = Sim(ncases=int(1e3))
    sim.run(n=2)
    return

if __name__ == '__main__':
    main()
```

**Anything else we need to know?**:

**Environment**:
Note that this behavior is not new and has been showing itself since I started using dask at the start of this year.

- Dask version: 2022.9.1 
- Python version: 3.10
- Operating System: Ubuntu running on Windows via WSL2
- Install method (conda, pip, source): pip

<details>
<summary>Cluster Dump State:</summary>

</details>
[dask-report.zip](https://github.com/dask/distributed/files/9641862/dask-report.zip)

**Comment by hayesgb** [comment]
The reason for the long delay between the two groups of tasks is the `dask.compute(*runcases)` call.  This blocks until the inner loop completes. The following alleviates that.

```
    def run(self, n):
        with performance_report():
            for _ in range(n):
                runcases = []
                for case in self.cases:
                    case_delayed = dask.delayed(case_runner)(case)
                    runcases.append(case_delayed)
            self.runcases = dask.compute(*runcases)
```
You can trigger computation without blocking with `dask.persist()`, or maybe take a look [here](https://docs.dask.org/en/stable/futures.html#waiting-on-futures) if you need to interleave submissions and computations.

**Comment by scottshambaugh** [comment]
I'm still confused, since even if `dask.compute()` is blocking, that inner loop should be complete at the 49 second mark in the plot above, right? Should it really take 5 whole seconds of wall time to gather all the results here when it's just a thousand ints? This is just an example, but the delay is scaling up in my real use.

**Comment by gjoseph92** [comment]
@scottshambaugh I'd suggest profiling this whole thing with [py-spy](https://github.com/benfred/py-spy): `py-spy record --idle --subprocesses --format=speedscope -o profile.json -- python repro.py`. That should give you a more definitive look at what's going on in the white space.

@hayesgb is correct that if you have a lot of things to compute, it's better to submit them all at once. But I also agree with you that this does feel slow.

> Should it really take 5 whole seconds of wall time to gather all the results here when it's just a thousand ints?

Profiling should help show what's going on.

You might also be interested in https://github.com/dask/distributed/issues/4754, which I have a feeling this might end up being a duplicate of.

The way `compute` currently works is that first, all the results have to be aggregated onto one worker. Then, that aggregated result is sent to the client. This means:
1. There's no overlap of communication and computation. Results don't start to move until the _last_ result has been computed. Ideally, you'd want each result to start moving to the client as soon as it's done. So when the last result completes, most of the others would have have gotten to the client already, or be on their way.
2. There's an extra hop of going to a worker first, before the client.

**Comment by mrocklin** [comment]
The delay could also be the new compute call, and graph submission time.
In my experience that is a more likely culprit than data transfer time.

**Comment by scottshambaugh** [comment]
Thank you all for the pointers! 

@gjoseph92 attached are that speedscope profiles which you can unzip to json and load with https://www.speedscope.app/ to examine. 

`profile_original_example.json` was run with my original code at the top of this post:
![image](https://user-images.githubusercontent.com/14363975/202921964-09797bd7-802b-4d37-b6af-6640c33c5d8a.png)

`profile_client_map.json` was run by substituting in the below block of code. It saves a good chunk of time not having the `gather` calls, though both it and the original run seem to spend a lot of time in `_get_computation_code` on the first call that I don't quite understand.
```python
def run(self, n):
    with performance_report():
        for _ in range(n):
            runcases = []
            self.runcases = self.client.map(case_runner, self.cases)
```
![image](https://user-images.githubusercontent.com/14363975/202921854-a5c288ae-a6ce-4b73-b755-473ec3a8ce0f.png)

For my purposes I think this is good enough of a speedup. I'll leave the issue open until someone reviews this and can confirm this is all expected behavior, at which point it can be closed.

[profiles.zip](https://github.com/dask/distributed/files/10050504/profiles.zip)
